src/acktest/bootstrapping/vpc.py

```diff
@@ -0,0 +1,185 @@
+        ec2_resource = boto3.resource("ec2", region_name=region)
+
+        vpc = ec2_client.create_vpc(CidrBlock=self.vpc_cidr_block)
+
+        self.vpc_id = vpc['Vpc']['VpcId']
+        logging.info(f"Created VPC {self.vpc_id}")
+
+        if self.name_prefix is not None:
+            self.name = resources.random_suffix_name(self.name_prefix, 63)
+            ec2_client.create_tags(Resources=[self.vpc_id], Tags=[{'Key': 'Name', 'Value': self.name}])
+
+        vpc = ec2_resource.Vpc(self.vpc_id)
+        vpc.wait_until_available()
+
+        self._create_internet_gateway()
+        public_route_table = self._create_public_route_table()
+        self.public_route_table_id = public_route_table.id
+
+        region_azs = self.get_availability_zone_names()
+
+        subnet_count = 0
+        for i in range(self.public_subnets):
+            if self.public_subnet_cidr_blocks is None:
+                cidr_block = f"10.0.{subnet_count}.0/24"
+            else:
+                cidr_block = self.public_subnet_cidr_blocks[i]
+            subnet = vpc.create_subnet(CidrBlock=cidr_block, AvailabilityZone=region_azs[subnet_count % len(region_azs)])
+            self.public_subnet_ids.append(subnet.id)
+
+            ec2_client.associate_route_table(RouteTableId=public_route_table.id, SubnetId=subnet.id)
+
+            logging.info(f"Created public subnet {subnet.id}")
+            subnet_count += 1
+
+        if self.private_subnets == 0:
+            return
+
+        private_route_table = self._create_private_route_table()
+        self.private_route_table_id = private_route_table.id
+
+        for i in range(self.private_subnets):
+            if self.private_subnet_cidr_blocks is None:
+                cidr_block = f"10.0.{subnet_count}.0/24"
+            else:
+                cidr_block = self.private_subnet_cidr_blocks[i]
+            subnet = vpc.create_subnet(CidrBlock=cidr_block, AvailabilityZone=region_azs[subnet_count % len(region_azs)])
+            self.private_subnet_ids.append(subnet.id)
+
+            ec2_client.associate_route_table(RouteTableId=private_route_table.id, SubnetId=subnet.id)
+
+            logging.info(f"Created private subnet {subnet.id}")
+            subnet_count += 1
+
+    def cleanup(self):
+        """Deletes a VPC.
+        """
+        region = get_region()
+        ec2 = boto3.client("ec2", region_name=region)
+        ec2_resource = boto3.resource("ec2", region_name=region)
+
+        vpc = ec2_resource.Vpc(self.vpc_id)
+
+        for subnet in self.public_subnet_ids + self.private_subnet_ids:
+            ec2.delete_subnet(SubnetId=subnet)
+
+            logging.info(f"Deleted subnet {self.name}")
+
+        ec2.delete_route_table(RouteTableId=self.private_route_table_id)
+        ec2.delete_route_table(RouteTableId=self.public_route_table_id)
+
+        vpc.detach_internet_gateway(InternetGatewayId=self.internet_gateway_id)
+        ec2.delete_internet_gateway(InternetGatewayId=self.internet_gateway_id)
```

Comment:
for these methods that create multiple resources, how are you handling errors and cleaning up resources created before an error occurs? For instance, if you successfully create the VPC, route table and public subnets but then hit some API quota or resource limit when creating private subnets, are you planning to clean up the previously created route tables, public subnets and VPC? Generally I'm not seeing any error handling at all for either `bootstrap` or `cleanup`. Is that on purpose?

Reply:
I'm rewriting VPC so it's composed of a number of smaller `Bootstrappable` types. I don't think how it's written at the moment can solve that issue.